osbuild/util/linux.py

```diff
@@ -274,3 +281,241 @@ def bits(n):
+    RENAME_WHITEOUT = ctypes.c_uint(4)
+
+    _lock = threading.Lock()
+    _inst = None
+
+    def __init__(self, lib: ctypes.CDLL):
+        self._lib = lib
+
+        # prototype: renameat2
+        proto = ctypes.CFUNCTYPE(
+            ctypes.c_int,
+            ctypes.c_int,
+            ctypes.c_char_p,
+            ctypes.c_int,
+            ctypes.c_char_p,
+            ctypes.c_uint,
+            use_errno=True,
+        )(
+            ("renameat2", self._lib),
+            (
+                (1, "olddirfd", self.AT_FDCWD),
+                (1, "oldpath"),
+                (1, "newdirfd", self.AT_FDCWD),
+                (1, "newpath"),
+                (1, "flags", 0),
+            ),
+        )
+        setattr(proto, "errcheck", self._errcheck_errno)
+        setattr(proto, "__name__", "renameat2")
+        self.renameat2 = proto
+
+    @staticmethod
+    def make() -> "Libc":
+        """Create a new instance"""
+
+        return Libc(ctypes.CDLL("", use_errno=True))
+
+    @classmethod
+    def default(cls) -> "Libc":
+        """Return and possibly create the default singleton instance"""
+
+        with cls._lock:
+            if cls._inst is None:
+                cls._inst = cls.make()
+            return cls._inst
+
+    @staticmethod
+    def _errcheck_errno(result, func, args):
+        if result < 0:
+            err = ctypes.get_errno()
+            msg = f"{func.__name__}{args} -> {result}: error ({err}): {os.strerror(err)}"
+            raise OSError(err, msg)
+        return result
+
+
+def proc_boot_id(appid: str):
+    """Acquire Application-specific Boot-ID
+
+    This queries the kernel for the boot-id of the running system. It then
+    calculates an application-specific boot-id by combining the kernel boot-id
+    with the provided application-id. This uses a cryptographic HMAC.
+    Therefore, the kernel boot-id will not be deducable from the output. This
+    allows the caller to use the resulting application specific boot-id for any
+    purpose they wish without exposing the confidential kernel boot-id.
+
+    This always returns an object of type `uuid.UUID` from the python standard
+    library. Furthermore, this always produces UUIDs of version 4 variant 1.
+
+    Parameters
+    ----------
+    appid
+        An arbitrary object (usually a string) that identifies the use-case of
+        the boot-id.
+    """
+
+    with open("/proc/sys/kernel/random/boot_id", "r", encoding="utf8") as f:
+        content = f.read().strip(" \t\r\n")
+
+    # Running the boot-id through HMAC-SHA256 guarantees that the original
+    # boot-id will not be exposed. Thus two IDs generated with this interface
+    # will not allow to deduce whether they share a common boot-id.
+    # From the result, we throw away everything but the lower 128bits and then
+    # turn it into a UUID version 4 variant 1.
+    h = bytearray(hmac.digest(content.encode(), appid.encode(), hashlib.sha256))  # type: ignore
```

Comment:
`hmac.digest` is a Python 3.7 addition, see [hmac.digest](https://docs.python.org/3/library/hmac.html#hmac.digest).

Or in our CI for the integration WIP PR:
```
rhel_87-aarch64-ami-boot Building & Comparing: FAIL: osbuild failed: Traceback (most recent call last):
[2351](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2351)  File "/usr/bin/osbuild", line 11, in <module>
[2352](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2352)    load_entry_point('osbuild==73', 'console_scripts', 'osbuild')()
[2353](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2353)  File "/usr/lib/python3.6/site-packages/osbuild/main_cli.py", line 152, in osbuild_cli
[2354](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2354)    with ObjectStore(args.store) as object_store:
[2355](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2355)  File "/usr/lib/python3.6/site-packages/osbuild/objectstore.py", line 468, in __enter__
[2356](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2356)    self._stack.enter_context(self.cache)
[2357](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2357)  File "/usr/lib64/python3.6/contextlib.py", line 330, in enter_context
[2358](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2358)    result = _cm_type.__enter__(cm)
[2359](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2359)  File "/usr/lib/python3.6/site-packages/osbuild/util/fscache.py", line 512, in __enter__
[2360](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2360)    self._bootid = linux.proc_boot_id(self._appid)
[2361](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2361)  File "/usr/lib/python3.6/site-packages/osbuild/util/linux.py", line 518, in proc_boot_id
[2362](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2362)    h = bytearray(hmac.digest(content.encode(), appid.encode(), hashlib.sha256))  # type: ignore
[2363](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild/-/jobs/3404205174#L2363)AttributeError: module 'hmac' has no attribute 'digest'
```